atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunner.java

```diff
@@ -15,19 +15,361 @@
+                    RangeRequests.getNextStartRow(false, rowResultTimestamps.lastItem().getRowName());
+            return SweepResults.builder()
+                    .previousStartRow(Optional.fromNullable(startRow))
+                    .nextStartRow(Optional.fromNullable(nextRow))
+                    .cellsExamined(rowResultTimestamps.size())
+                    .cellsDeleted(totalCellsSwept.get())
+                    .sweptTimestamp(sweepTs)
+                    .build();
+        }
+    }
+
+    private ClosableIterator<RowResult<Value>> getValues(TableReference tableRef,
+                                                         RangeRequest range,
+                                                         long sweepTs,
+                                                         Sweeper sweeper) {
+        if (sweeper.shouldSweepLastCommitted()) {
+            return keyValueService.getRange(tableRef, range, sweepTs);
+        } else {
+            return ClosableIterators.emptyImmutableClosableIterator();
+        }
+    }
+
+    public static Equivalence<RowResult<Set<Long>>> sameRowEquivalence() {
+        return new Equivalence<RowResult<Set<Long>>>() {
+            @Override
+            protected boolean doEquivalent(RowResult<Set<Long>> fst, RowResult<Set<Long>> snd) {
+                return Arrays.equals(fst.getRowName(), snd.getRowName());
+            }
+
+            @Override
+            protected int doHash(RowResult<Set<Long>> setRowResult) {
+                return 0;
+            }
+        };
+    }
+
+    private int sweepForCells(
+            CellsAndTimestamps currentBatch,
+            TableReference tableRef,
+            Sweeper sweeper,
+            long sweepTs,
+            PeekingIterator<RowResult<Value>> peekingValues,
+            RunType runType) {
+        CellsAndTimestamps currentBatchWithoutIgnoredTimestamps =
+                currentBatch.withoutIgnoredTimestamps(sweeper.getTimestampsToIgnore());
+
+        CellsToSweep cellsToSweep = getStartTimestampsPerRowToSweep(
+                currentBatchWithoutIgnoredTimestamps, peekingValues, sweepTs, sweeper);
+
+        Multimap<Cell, Long> startTimestampsToSweepPerCell = cellsToSweep.timestampsAsMultimap();
+
+        if (runType == RunType.FULL) {
+            cellsSweeper.sweepCells(tableRef, startTimestampsToSweepPerCell, cellsToSweep.allSentinels());
+        }
+
+        return startTimestampsToSweepPerCell.size();
+    }
+
+    private static Iterator<CellAndTimestamps> getTimestampsFromRowResultsIterator(
+            Iterable<RowResult<Set<Long>>> cellsToSweep) {
+        return StreamSupport.stream(cellsToSweep.spliterator(), false)
+                .flatMap(SweepTaskRunner::rowToCellAndTimestampStream)
+                .iterator();
+    }
+
+    private static Stream<CellAndTimestamps> rowToCellAndTimestampStream(RowResult<Set<Long>> rowResult) {
+        Set<Map.Entry<Cell, Set<Long>>> cellsInRow = ImmutableSet.copyOf(rowResult.getCells());
```

Comment:
Is the set copy here to allow the row to be GCed? Could you just stream the entries?